### ISSUE TITLE
Metadata not always emitting when federating

### ISSUE THREAD

**Issue by pheyvaer** [comunica/comunica#553]
#### Issue type:

- :bug: Bug

____
#### Description:

I execute a GraphQL-LD query on 8 sources via `actor-init-sparql` that works, but when adding a 9th one it stops before providing any results.

The query is
```
{
  id(_:KNOWS)
  employee {
    givenName @single
    familyName @single
    img @single
    title @single
  }
}

```
The context is 
```
{
    "foaf": "http://xmlns.com/foaf/0.1/",
    "vcard": "http://www.w3.org/2006/vcard/ns#",
    "schema": "http://schema.org/",
    "givenName": {
      "@id": "foaf:givenName",
      "@language": "en"
    },
    "familyName": {
      "@id": "foaf:familyName",
      "@language": "en"
    },
    "img": {
      "@id": "foaf:img"
    },
    "employee": "schema:employee",
    "title": "vcard:title",
    "KNOWS": "http://knows.proxy.ilabt.imec.be/person/office/#"
  }
```

Sources that work
```
https://ruben.verborgh.org/profile/#me  
http://knows.proxy.ilabt.imec.be/person/erikmannens/#me 
https://pieterheyvaert.com/#me 
http://knows.proxy.ilabt.imec.be/person/office/# 
https://ben.de-meester.org/#me https://www.rubensworks.net/#me 
https://brechtvdv.solid.community/profile/card#me 
https://sven-lieber.org/profile#me 
http://knows.proxy.ilabt.imec.be/person/geraldh/#me 
```
Sources that fail (one added to the previous list)
```
https://ruben.verborgh.org/profile/#me  
http://knows.proxy.ilabt.imec.be/person/erikmannens/#me 
https://pieterheyvaert.com/#me 
http://knows.proxy.ilabt.imec.be/person/office/# 
https://ben.de-meester.org/#me https://www.rubensworks.net/#me 
https://brechtvdv.solid.community/profile/card#me 
https://sven-lieber.org/profile#me 
http://knows.proxy.ilabt.imec.be/person/geraldh/#me 
http://knows.proxy.ilabt.imec.be/person/harm/#me
```
____
#### Environment:

| software            | version
| ------------------- | -------
| Comunica Init Actor | 1.9.2 (dev)
| node                | v11.3.0
| npm                 | 6.10.3
| yarn                | 1.16.0
| Operating System    | linux (Linux 4.15.0-65-generic)

**Comment by rubensworks** [comment]
I highly doubt the number of sources are causing the problem.

What happens when you re-order the sources?
Do you see any specific console errors/warnings?

**Comment by pheyvaer** [comment]
Reordering the sources doesn't have an effect. When the using 9 sources the actor just stops without any output (when the log level is the default one).

**Comment by RubenVerborgh** [comment]
8 is an interesting number of course; I wonder if some buffer is being set to 8.

**Comment by rubensworks** [comment]
True, that may be possible.

AFAIK, most of the asynciterator buffers either make use of buffersizes of 4 or 128, but there may be one of size 8 that I'm unaware of.

**Comment by joachimvh** [comment]
I'm currently looking into this. The issue is that it doesn't work with 10 sources and it does work with 9 because if you look closely to Pieter's list of sources there is 1 line that contains 2 sources :D.

Also a simple select all query does give results, while this specific query here doesn't even give an empty result (it never reaches the point of returning the result output). Not yet sure what's up with that.

**Comment by rubensworks** [comment]
Perhaps a promise that is never being resolved somewhere?
Could be because of a source that hangs?

**Comment by RubenVerborgh** [comment]
It’s not the duplicates, is it?

**Comment by joachimvh** [comment]
So this is an annoying one to track down due to all the promises and lazy stuff.

Interesting finds so far:
 - There is a problem [here](https://github.com/comunica/comunica/blob/master/packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts#L149-L193) in the federated actor. When extracting metadata for the first pattern everything is fine, when doing it for the second pattern for some reason only 9 of the 10 created PromiseProxyIterators get called. Since the last one never gets called, the trigger to fire the metadata result never happens (or something causes to code to stop before the 10 call could happen).
 - The problem is fixed if the buffer in ActorRdfResolveQuadPatternHypermedia is disabled. No idea why but this makes me think there might be a timing problem. This also explains why the problem only happens for the second metadata call (since the caching happens in the first).
 - Enabling the cache, but setting the maxBufferSize of the RoundRobinUnionIterator to 8 or higher (and autoStart to false, took me quite a while before I realized that was also required :D ) also fixes the issue (no idea why the minimum is 8). So again indicating a potential timing problem (or who knows, even a bug outside of Comunica).

But so no idea yet why this happens since there are many locations that can be the cause of this.

**Comment by RubenVerborgh** [comment]
Good sleuthing, very interesting…

Regarding the first piece of code you pointed to, a worry I have there is the `remainingSources` variable that serves a double purpose. Perhaps would be good to have a separate `metadataEmitted` variable to avoid the "setting it to 0" trick. Likely not related though.

Given your above comments, a timing problem is indeed most likely. I wonder if there is some catch-22 that we are ending up in, where iterators are essentially waiting for each other to continue, whereas with the buffer size change, more things happen in parallel.

**Comment by mielvds** [comment]
I remember back in the day we had the same issue with the federation mediator in Client.js. Some iterators were not emitting the metadata event for some reason. Unfortunately I don't recall exactly what the issue was. Maybe this is a lead: https://github.com/LinkedDataFragments/Client.js/commit/46df9aedaff2019dbcd21d3ae6f62389254817b0

**Comment by mielvds** [comment]
Also https://github.com/LinkedDataFragments/Client.js/commit/0f4024f9b5e273304366e7e203d27539c80afe2e and https://github.com/LinkedDataFragments/Client.js/commit/ce2866c14462d9c657a1f6728e685fbd402247ca

**Comment by RubenVerborgh** [comment]
There's definitely a general pattern to be learned here, apart from the specific bug.

The assumption seems to be that callbacks always return / promises always resolve. Shall we incorporate a pattern where they time out? So consistently, when we wait for a promise, surround them with an `expires` construct or so. Where this function is controlled by some parameters (duration, whether it is a fatal error or a warning).

**Comment by mielvds** [comment]
Like this: https://www.npmjs.com/package/nodejs-promise-timeout? That is dangerous though, how do you prevent blocking slower operations? Maybe a progress callback as means for debugging is a compromise?

**Comment by RubenVerborgh** [comment]
> That is dangerous though, how do you prevent blocking slower operations?

By making it configurable globally, such that the timeout can be infinity, or that they cause hard errors.

**Comment by joachimvh** [comment]
So I'm actually still not sure exactly why this happens. I do think it can be fixed by changing some of the code so it's not dependent on data being read from somewhere else before the metadata can be started, which I'll do tomorrow because at this point I'm giving up at pinpointing exactly why this happens :D.

Some more tidbits:
 - The bug also happens with 11-13 sources, once you have 14+ sources though it works again!
 - Replacing the fillbuffer call on round robin iterator with a call draining the data from that iterator also makes sure all the metadata gets generated (but of course fails because the data disappeared), showing again that there is an issue with the read function not being called anymore.
 - I actually even have no idea why this code works for 5+ sources in the first place, because fillbuffer would trigger the first 4 sources but as far as I can see there is nothing actually listening to the data stream so I don't even know why the 5th source is being read at some point. Which is another reason why I'm just going to implement a (hopefully) fix even though I can't find the exact reason this happens.

**Comment by RubenVerborgh** [comment]
>     * Replacing the fillbuffer call on round robin iterator with a call draining the data from that iterator also makes sure all the metadata gets generated (but of course fails because the data disappeared)

We probably want to replace that mechanism with a proper splitter, that extracts metadata in one stream (which is pulled, so it will drain) and data in another (which is not pulled initially).

Relying on side-effects of `_fillBuffer` to make the above happen is not reliable.

**Comment by rubensworks** [comment]
We actually have these two streams since the last refactoring.

However, on a higher level, this metadata streams is collected into a single metadata object. And it's this object that is emitted as a metadata-event in the main data stream.

I haven't checked if this is possible, but perhaps we can remove this metadata-event, and instead pass it as a callback-wrapped promise next to the data stream.

**Comment by RubenVerborgh** [comment]
> I haven't checked if this is possible, but perhaps we can remove this metadata-event, and instead pass it as a callback-wrapped promise next to the data stream.

I think so, yes. The `metadata` event design originates from Client.js, and there it originates from the precursor to AsyncIterator. Probably would not design it like that anymore if I were to do it again.

**Comment by joachimvh** [comment]
Current suggested solution that doesn't require restructuring the entire flow (because I don't really know where to even start there :D ):

So the issue is that [this](https://github.com/comunica/comunica/blob/master/packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts#L173-L184) part doesn't get triggered for all sources. Reason being that due to how PromiseProxyIterators are made, they only call that function with their first read call, which doesn't happen because timing issue somewhere between all the nested _fillBuffer calls and the cache (stacktracke contains like 3+ nested _fillBuffer calls at some points so pinpointing it is a bit hard) probably causes all buffers in the BufferedIterators to fill before that one gets called.

The cleanest solution I currently have is changing the PromiseProxyIterator class so the sources can be loaded without having to depend on a read call. That means moving [this](https://github.com/rubensworks/asynciterator-promiseproxy.js/blob/master/lib/PromiseProxyIterator.ts#L19-L23) to a separate function that can be called externally. In the FederatedQuadSource class the stream of those iterators would then be cloned and that function would be called on all entries once the metadata call occurs (so that instead of the _fillBuffer call). That would cause all metadata calls to happen (I tested this and it works). It only requires a few lines of changes.

Besides that I'm sort of out of ideas. There are some others things I tried to just keep it in Comunica but I never got those working because the promises sometimes don't fire (and I have no idea why).

**Comment by RubenVerborgh** [comment]
> It only requires a few lines of changes.

Sounds good to try then 🙂 Could you PR it?

**Comment by rubensworks** [comment]
@joachimvh Not sure if I fully understand your suggestion solution, but it sounds like this would make it non-lazy. Also, the cloning will probably introduce a performance hit. So I'm not really convinced of this solution.

> So the issue is that this part doesn't get triggered for all sources. Reason being that due to how PromiseProxyIterators are made, they only call that function with their first read call, which doesn't happen because timing issue somewhere between all the nested _fillBuffer calls and the cache (stacktracke contains like 3+ nested _fillBuffer calls at some points so pinpointing it is a bit hard) probably causes all buffers in the BufferedIterators to fill before that one gets called.

It makes sense that the cloning solution solves this problem. But it feels like a cheap bandaid fix for a deeper problem, which we should properly resolve IMO. It's possible that it's just the metadata event from one of the sources that is being emitted too late somehow.

**Comment by mielvds** [comment]
> It's possible that it's just the metadata event from one of the sources that is being emitted too late somehow.

In what scenarios could that occur? This might bring us closer

**Comment by joachimvh** [comment]
> Not sure if I fully understand your suggestion solution, but it sounds like this would make it non-lazy. Also, the cloning will probably introduce a performance hit.

It would stay lazy, the call to the PromiseProxyIterators would happen at the same time the _fillBuffer call happens now (when a metadata listener is added). Also the clone is of the source iterator (to make sure it stays lazy) so would only be of a few elements. It's definitely less of a bandaid fix than the previous solution since now we say what we want (load the sources so that they do their magic and start parsing the metadata) instead of the current _fillBuffer call (start reading some data just so the read function of some classes gets called which would cause the metadata parsing to start).

>It's possible that it's just the metadata event from one of the sources that is being emitted too late somehow.

Currently the problem is that the `read` function of the 10th PromiseProxyIterator is never called, causing it to never call its async source function. And it is this function that is necessary to start the metadata parsing, hence the suggestion of calling this function ourselves since that is what we want to happen anyway.

I can't say exactly why it never gets called though. I think it has something to do with all the buffers being full everywhere (since nobody is actually reading those bindings from the RoundRobinUnionIterator, they just got put in buffers there to force the read calls on the PromiseProxyIterators).

**Comment by rubensworks** [comment]
Not sure I fully understand your intended solution. Just PR'ing it may be easier to understand. We can further discuss this based on the PR.

**Comment by joachimvh** [comment]
The change suggested requires rubensworks/asynciterator-promiseproxy.js#7 . In Comunica itself [this](https://github.com/comunica/comunica/blob/master/packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts#L149-L198) block would then be replaced with

```typescript
    const proxyIt: RoundRobinUnionIterator<RDF.Quad> = sourcesIt.map((source) => {
      // ...
      });
    })
    const it = new RoundRobinUnionIterator(proxyIt.clone());
    it.on('newListener', (eventName) => {
      if (eventName === 'metadata') {
        setImmediate(() => proxyIt.clone().each((proxy) => proxy.loadSource());
      }
    });
```

**Comment by rubensworks** [comment]
I still feel like there's too much cloning going on. For this issue, there are just 8 sources, so this won't cause an issue here. However, when we start moving towards many-source use cases (which will be the case with Solid and link traversal), we may run into some performance issues due to this.

I haven't looked into this issue in detail yet, so I don't have a good suggestion on alternative quick solutions, but I don't think cloning is the way to go.

**Comment by RubenVerborgh** [comment]
> I still feel like there's too much cloning going on.

Just want to mention that, in `asynciterator`, cloning is relatively lightweight. Yes, it does create a buffer of the items that are read so far, but that buffer is shared among clones. So cloning something once versus multiple times does not make a significant difference.

But your objections to cloning might be different.

**Comment by joachimvh** [comment]
It's the easiest way I found to fix the problem while still a) keeping the source iterator lazy, and b) keeping the data pulling lazy. There are definitely solutions without cloning but those would probably require more drastic restructuring.

**Comment by rubensworks** [comment]
> So cloning something once versus multiple times does not make a significant difference.

Ok, good to know. In any case, the buffer is still kept, so for many/unbounded sources, we may run into problems in the future.

> There are definitely solutions without cloning but those would probably require more drastic restructuring.

I acknowledge that this is a problem *now*, and that we currently can not afford to spent a lot of time refactoring stuff to fix this the proper way.

So I agree to move forward with the suggested fix, but include a big `TODO` describing the problems of this fix, and opening an accompanying issue (with major priority) to fix this asap. We/I may have some time to look into this early next year.

**Comment by rubensworks** [comment]
While I'm not 100% certain, I think I just saw this issue popping up again for the following query on http://ldf.nanopubs.lod.labs.vu.nl/np:
```
prefix np: <http://www.nanopub.org/nschema#>
prefix npx: <http://purl.org/nanopub/x/>
prefix npa: <http://purl.org/nanopub/admin/>
prefix dct: <http://purl.org/dc/terms/>
prefix foaf: <http://xmlns.com/foaf/0.1/>
select ?user ?name ?intronp ?date ?pubkey where {
  graph npa:graph {
    ?intronp npa:hasHeadGraph ?h .
    ?intronp npa:hasValidSignatureForPublicKey ?pubkey .
    ?intronp dct:created ?date .
  }
  graph ?h {
    ?intronp np:hasAssertion ?a .
  }
  graph ?a {
    ?keydeclaration npx:declaredBy ?user .
    optional {
      ?user foaf:name ?name .
    }
  }
}
```

When removing the optional block, things seem to work again.

I have not been able to look into it in detail yet, but adding [this line](https://github.com/comunica/comunica/pull/632/files#diff-7ca54ff1bb9203da2105dc6f910d8c61R146) in the LeftJoin actor seems to resolve the problem. This indicates that the metadata event must be emitted too early in the federated actor, which causes the metadata promise to never resolve, since it is waiting on an event that has already been emitted before.

This makes me wonder why we are using events instead of AsyncIterator properties...

In any case, the current metadata system is a bit messy, and deserves a rewrite.

**Comment by rubensworks** [comment]
Root cause: https://github.com/RubenVerborgh/AsyncIterator/issues/22

**Comment by rubensworks** [comment]
Fixed in #733
